gustaf/ffd.py

```diff
@@ -1,4 +1,579 @@
+    def spline(self, spline: SPLINE_TYPES):
+        """
+        Sets spline. The spline parametric range bounds will be converted 
+        into the bounds [0,1]^para_dim.
+
+        Parameters
+        -----------
+        spline: SPLINE_TYPES
+            New Spline for the next deformation
+
+        Returns
+        --------
+        None
+        """
+        self._spline = spline.copy()
+        self._scale_parametric_dimension_to_hypercube()
+
+        self._logi("Setting Spline.")
+        self._logi("Spline Info:")
+        self._logi(
+            f"  Parametric dimensions: {spline.para_dim}."
+        )
+        self.calculated = False
+
+    @property
+    def is_partial(self):
+        """
+        Returns is_partial.
+
+        Parameters
+        -----------
+        None
+
+        Returns
+        --------
+        self._is_partial: bool
+        """
+        if self._is_partial:
+            self._logw("Partial FFD is currently not supported.")
+        return self._is_partial
+
+    @is_partial.setter
+    def is_partial(self, is_partial):
+        """
+        Setter for is_partial. Do you want partial FFD of your mesh?
+
+        Parameters
+        -----------
+        is_partial: bool
+
+        Returns
+        --------
+        None
+        """
+        if is_partial:
+            self._logw("Partial FFD is currently not supported.")
+        # self._is_partial = is_partial
+        # self._logd("This FFD will be:")
+        # if is_partial:
+        #     self._logw("  Partial!")
+        #     self._logw("  Be careful! Else, mesh might tangle!")
+        #     self._offset = None
+        #     self._scale = None
+
+        # else:
+        #     self._logd("  Non-partial!")
+
+
+    def _scale_parametric_dimension_to_hypercube(self):
+        """Scales all knot_vectors of the spline to a range of [0,1]
+        """
+        if self._is_parametric_room_hypercube(self._spline):
+            pass
+        else:
+            knv = []
+            for knot_vector in self._spline.knot_vectors:
+                knot_vector = [kv_v - knot_vector[0] for kv_v in knot_vector]
+                knot_vector = [kv_v / knot_vector[-1] for kv_v in knot_vector]
+                knv.append(knot_vector)
+            self._spline.knot_vectors = knv
+
+
+    def _is_parametric_room_hypercube(
+            self,
+            spline: SPLINE_TYPES
+        ) -> bool:
+        """Checks if the parametric range of the spline is a hypercube.
+
+        Parameters
+        ----------
+        spline : SPLINE_TYPES
+            New spline for which to check for hypercubeness
+
+        Returns
+        -------
+        bool
+            True if parametric room is hypercube, else False
+        """
+        if "Bezier" not in spline.whatami:
+            for knot_vector in spline.knot_vectors:
+                # check if knot_vectors fist element is 0 and last element is 1
+                if not (knot_vector[0] == 0 and knot_vector[-1] == 1):
```

Comment:
probably should be a floating point comparison. you can use global value `settings.TOLERANCE`